var/spack/repos/builtin/packages/adios2/package.py

```diff
@@ -51,11 +51,19 @@ class Adios2(CMakePackage):
             description='Enable the SST staging engine')
     variant('dataman', default=True,
             description='Enable the DataMan engine for WAN transports')
+    variant('dataspaces', default=False,
+            description='Enable support for DATASPACES')
     variant('ssc', default=True,
             description='Enable the SSC staging engine')
     variant('hdf5', default=False,
             description='Enable the HDF5 engine')
 
+    # transitive dependencies with optional dependencies
+    variant('libffi', default=True,
```

Comment:
We shouldn't have ffi and fabric as varaints, jsut make the dependencies conditional on `+sst`

Reply:
But libffi is used from DILL, is DILL a direct SST dependency (and only a SST dependency)?

Reply:
Found this in `cmake/adios2-config.cmake.in` which suggests so:
```cmake
if(@ADIOS2_HAVE_SST@)
  set(EVPath_DIR @EVPath_DIR@)
  if(NOT @BUILD_SHARED_LIBS@)
    set(atl_DIR @atl_DIR@)
    set(dill_DIR @dill_DIR@)
    set(ffs_DIR @ffs_DIR@)
    set(enet_DIR @enet_DIR@)
  endif()
endif()
```

Reply:
we could also expose some of the SST transitive dependencies in a multi-variant, which will make user-level disabling of those optional dependencies easier in case they do not build on some systems.

Reply:
As I understand it, sst in adios directly depends on libfabric, evpath, atl.  evpath then in turn depends on atl, libfabric, and ffs.  ffs then depends on dill.  Dill then depends on ffi.  So the dependency tree looks like:

- adios2(sst)
  - nvstream
  - craydrc
  - libfabric
  - evpath
    - craydrc
    - nvstream
    - libfabric
    - atl
    - ffs
      - dill
        - ffi
        - flex
        - bison
  - atl

Some of these are optional, some are not; @eisenhauer do I have that right?  I'm hesitant to really encode the full set of build flags as variants since I think it results in a needlessly complex package and will create difficulty for upstream dependees in the concretization.  I'd rather support a smaller number of fixed configurations that turn on or off sets of build flags in adios.

Reply:
Some changes to your list, moving flex,bison directly under FFS and adding notes.

adios2(sst)
- nvstream (for optional/experimental NVstream transport)
- libfabric (if not present RDMA transport will not be built, but sockets-only libfabric is useless)
- craydrc   (used by RDMA transport on Crays)
- evpath  
   - libfabric (leaving it in this list , but the EVPATH_NO_RDMA in adios2 build means evpath wont use it)
   - atl 
   - ffs
      - flex (optional, if our prebuilt source suffices)
      - bison (optional, if our prebuilt source suffices)
      - dill
        - ffi (we're always looking for this currently, but should build/run fine if it's not provided)
- atl (used directly also)
- ffs (used directly also)

notes:
- libfabric is a wrapper around whatever native RDMA exists on the platform, so where it is useful, it depends upon system libraries like verbs, rdma, etc.  
- there is currently no distribution for nvstream
- ffi *should* be unnecessary everywhere, but some platforms (where we have no native DCG in dill) might have build problems without it, unless we tweak some things.

Reply:
Thank you both for the details, that means the current PR gets things right and is ready.

> I'm hesitant to really encode the full set of build flags as variants since I think it results in a needlessly complex package and will create difficulty for upstream dependees in the concretization. I'd rather support a smaller number of fixed configurations that turn on or off sets of build flags in adios.

Yes, either we make big, bold block like now (with our without SST, as soon as something does not build; deactivate some functionality to avoid system pickup when found elsewhere) or we expose some dependencies in their own packages, which have their own control if needed. So far, approach 1 with the dependencies outlined above works for me. :+1: